library/src/main/java/com/datatorrent/lib/state/managed/ManagedTimeStateImpl.java

```diff
@@ -0,0 +1,125 @@
+
+import java.util.concurrent.Future;
+
+import javax.validation.constraints.Min;
+
+import com.google.common.util.concurrent.Futures;
+
+import com.datatorrent.api.annotation.OperatorAnnotation;
+import com.datatorrent.lib.state.BucketedState;
+import com.datatorrent.lib.state.TimeSlicedBucketedState;
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * This implementation of {@link ManagedState} lets the client to specify the time for each key. The value of time
+ * is used to derive the time-bucket of a key.
+ */
+@OperatorAnnotation(checkpointableWithinAppWindow = false)
+public class ManagedTimeStateImpl extends AbstractManagedStateImpl implements TimeSlicedBucketedState
+{
+  public ManagedTimeStateImpl()
+  {
+    this.numBuckets = 1;
+  }
+
+  @Override
+  public void put(long bucketId, long time, Slice key, Slice value)
+  {
+    int bucketIdx = prepareBucket(bucketId);
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket != -1) {
+      buckets[bucketIdx].put(key, timeBucket, value);
+    }
+  }
+
+  @Override
+  public Slice getSync(long bucketId, Slice key)
+  {
+    int bucketIdx = prepareBucket(bucketId);
+    return buckets[bucketIdx].get(key, -1, Bucket.ReadSource.ALL);
+  }
+
+  @Override
+  public Slice getSync(long bucketId, long time, Slice key)
+  {
+    int bucketIdx = prepareBucket(bucketId);
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket == -1) {
+      //time is expired so no point in looking further.
+      return BucketedState.EXPIRED;
+    }
+    return buckets[bucketIdx].get(key, timeBucket, Bucket.ReadSource.ALL);
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public Future<Slice> getAsync(long bucketId, Slice key)
+  {
+    int bucketIdx = prepareBucket(bucketId);
+    Bucket bucket = buckets[bucketIdx];
+    synchronized (bucket) {
+      Slice cachedVal = buckets[bucketIdx].get(key, -1, Bucket.ReadSource.MEMORY);
+      if (cachedVal != null) {
+        return Futures.immediateFuture(cachedVal);
+      }
+      return readerService.submit(new KeyFetchTask(bucket, key, -1, throwable));
+    }
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public Future<Slice> getAsync(long bucketId, long time, Slice key)
+  {
+    int bucketIdx = prepareBucket(bucketId);
+    Bucket bucket = buckets[bucketIdx];
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket == -1) {
+      //time is expired so no point in looking further.
+      return Futures.immediateFuture(BucketedState.EXPIRED);
+    }
+    synchronized (bucket) {
+      Slice cachedVal = buckets[bucketIdx].get(key, timeBucket, Bucket.ReadSource.MEMORY);
+      if (cachedVal != null) {
+        return Futures.immediateFuture(cachedVal);
+      }
+      return readerService.submit(new KeyFetchTask(bucket, key, timeBucket, throwable));
+    }
```

Comment:
This synchronized block and the assignment to `bucket` are duplicated in multiple places (3 or 4); could it be factored out to a helper method in the base class ?